src/borg/_hashindex.c

```diff
@@ -376,31 +377,69 @@ hashindex_get(HashIndex *index, const void *key)
     return BUCKET_ADDR(index, idx) + index->key_size;
 }
 
+inline
+int
+distance(HashIndex  *index, int current_idx, int ideal_idx)
+{
+    /* If the current index is smaller than the ideal index we've wrapped
+       around the end of the bucket array and need to compensate for that. */
+    return current_idx - ideal_idx + ( (current_idx < ideal_idx) ? index->num_buckets : 0 );
+}
+
 static int
 hashindex_set(HashIndex *index, const void *key, const void *value)
 {
     int idx = hashindex_lookup(index, key);
-    uint8_t *ptr;
+    uint8_t *bucket_ptr;
+    int offset = 0;
+    int other_offset;
+    void *bucket = malloc(index->key_size + index->value_size);
+    void *buffer = malloc(index->key_size + index->value_size);
```

Comment:
it would be cool if we could somehow not allocate/free bucket and buffer once per hashindex_set operation.

maybe the index data structure could have 2 pointers to such buffers that are just allocated once?